core/src/main/java/com/datastax/oss/driver/internal/core/cql/DefaultPagingState.java

```diff
@@ -0,0 +1,159 @@
+      ByteBuffer rawPagingState, Statement<?> statement, AttachmentPoint attachmentPoint) {
+    this(
+        rawPagingState,
+        hash(statement, rawPagingState, attachmentPoint),
+        attachmentPoint.getProtocolVersion().getCode());
+  }
+
+  private DefaultPagingState(ByteBuffer rawPagingState, byte[] hash, int protocolVersion) {
+    this.rawPagingState = rawPagingState;
+    this.hash = hash;
+    this.protocolVersion = protocolVersion;
+  }
+
+  // Same serialized form as in driver 3:
+  //     size of raw state|size of hash|raw state|hash|protocol version
+  //
+  // The protocol version might be absent, in which case it defaults to V2 (this is for backward
+  // compatibility with 2.0.10 where it is always absent).
+  public static DefaultPagingState fromBytes(byte[] bytes) {
+    ByteBuffer buffer = ByteBuffer.wrap(bytes);
+    short rawPagingStateLength = buffer.getShort();
+    short hashLength = buffer.getShort();
+    if (rawPagingStateLength + hashLength != buffer.remaining()
+        && rawPagingStateLength + hashLength + 2 != buffer.remaining()) {
+      throw new IllegalArgumentException(
+          "Cannot deserialize paging state, invalid format. "
+              + "The serialized form was corrupted, or not initially generated from a PagingState object.");
+    }
+    byte[] rawPagingState = new byte[rawPagingStateLength];
+    buffer.get(rawPagingState);
+    byte[] hash = new byte[hashLength];
+    buffer.get(hash);
+    int protocolVersion = buffer.hasRemaining() ? buffer.getShort() : 2;
+    return new DefaultPagingState(ByteBuffer.wrap(rawPagingState), hash, protocolVersion);
+  }
+
+  @Override
+  public byte[] toBytes() {
+    ByteBuffer buffer = ByteBuffer.allocate(rawPagingState.remaining() + hash.length + 6);
+    buffer.putShort((short) rawPagingState.remaining());
+    buffer.putShort((short) hash.length);
+    buffer.put(rawPagingState.duplicate());
+    buffer.put(hash);
+    buffer.putShort((short) protocolVersion);
+    buffer.rewind();
+    return buffer.array();
+  }
+
+  public static DefaultPagingState fromString(String string) {
+    byte[] bytes = Bytes.getArray(Bytes.fromHexString("0x" + string));
+    return fromBytes(bytes);
+  }
+
+  @Override
+  public String toString() {
+    return Bytes.toHexString(toBytes()).substring(2); // remove "0x" prefix
+  }
+
+  @Override
+  public boolean matches(@NonNull Statement<?> statement, @Nullable Session session) {
+    AttachmentPoint attachmentPoint =
+        (session == null) ? AttachmentPoint.NONE : session.getContext();
+    byte[] actual = hash(statement, rawPagingState, attachmentPoint);
+    return Arrays.equals(actual, hash);
+  }
+
+  @NonNull
+  @Override
+  public ByteBuffer getRawPagingState() {
+    return rawPagingState;
+  }
+
+  // Hashes a statement's query string and parameters. In retrospect we probably didn't need to
+  // include the paging state in the hash, but keeping it to have exactly the same output as driver
```

Comment:
> we probably didn't need to include the paging state in the hash
Actually I think I remember now, it provides some additional protection if a malicious actor tries to forge a serialized form. If the state is not included in the hash, you can send replace it with anything and it will get accepted.
I'll amend the comment.